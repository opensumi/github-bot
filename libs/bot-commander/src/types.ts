export type CompareFunc<T> = (command: T, userInput: string) => boolean;

export type FuncName = 'equal' | 'startwiths';

export interface ITextResolveResult {
  handler: any;
  type: 'text';
  /**
   * removed prefix
   */
  command: string;
}

export interface IRegexResolveResult {
  type: 'regex';
  regex: RegExp;
  handler: any;
  result: RegExpExecArray;
  command: string;
}

export interface IStarResolveResult {
  handler: any[];
  type: 'star';
  /**
   * removed prefix
   */
  command: string;
}

export type IResolveResult =
  | ITextResolveResult
  | IRegexResolveResult
  | IStarResolveResult;
