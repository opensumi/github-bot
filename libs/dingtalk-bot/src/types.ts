import type { RobotTextMessage } from 'dingtalk-stream';

export interface TextMessage {
  content: string;
}

export interface AtUsersItem {
  dingtalkId: string;
  staffId?: string;
}

export enum ConversationType {
  private = '1',
  group = '2',
}

export interface Message extends RobotTextMessage {
  conversationType: ConversationType | string;
}

export interface PrivateMessage extends Message {
  conversationType: ConversationType.private;
}

export interface GroupMessage extends Message {
  atUsers: AtUsersItem[];
  conversationType: ConversationType.group;
  conversationTitle: string;
  isInAtList: boolean;
}

export function isPrivateMessage(message: any): message is PrivateMessage {
  return (
    message &&
    message.conversationType &&
    message.conversationType === ConversationType.private
  );
}

export function isGroupMessage(message: any): message is GroupMessage {
  return (
    message &&
    message.conversationType &&
    message.conversationType === ConversationType.group
  );
}

export interface At {
  at: {
    atDingtalkIds: string[];
  };
}

export type Base<T extends string, O> = Record<T, O> & { msgtype: T };

export type Image = Base<
  'image',
  {
    picURL: string;
  }
>;

export type Markdown = Base<
  'markdown',
  {
    title: string;
    text: string;
  }
>;

export type Text = Base<
  'text',
  {
    content: string;
  }
>;

export type Link = Base<
  'link',
  {
    title: string;
    text: string;
    messageUrl: string;
    picUrl: string;
  }
>;

export type SingleActionCard = Base<
  'actionCard',
  {
    /**
     * 首屏会话透出的展示内容。
     */
    title: string;
    /**
     * markdown格式的消息。
     */
    text: string;

    singleTitle: string;
    singleURL: string;
  }
>;

export type MultiActionCard = Base<
  'actionCard',
  {
    /**
     * 首屏会话透出的展示内容。
     */
    title: string;
    /**
     * markdown格式的消息。
     */
    text: string;
    btns: {
      title: string;
      actionURL: string;
    }[];
    /**
     * 0：按钮竖直排列
     * 1：按钮横向排列
     */
    btnOrientation?: '0' | '1';
  }
>;

export type SendMessage =
  | Image
  | Markdown
  | Text
  | Link
  | SingleActionCard
  | MultiActionCard;

export function atDingtalkIds(...atDingtalkIds: string[]): At {
  return {
    at: {
      atDingtalkIds,
    },
  };
}

export function text(content: string): Text {
  return {
    msgtype: 'text',
    text: {
      content: content ? content.toString().trim() : '',
    },
  };
}

export function compose(...objects: any[]) {
  return Object.assign({}, ...objects);
}

export function markdown(title: string, text: string): Markdown {
  return {
    msgtype: 'markdown',
    markdown: {
      title,
      text,
    },
  };
}

export function image(url: string): Image {
  return {
    msgtype: 'image',
    image: { picURL: url },
  };
}

export function extension(extension: Record<string, unknown>) {
  return {
    msgtype: 'extension',
    extension,
  };
}

export function code(language: string, code: string) {
  return {
    ...extension({ text_type: 'code_snippet', code_language: language }),
    ...text(code),
  };
}
