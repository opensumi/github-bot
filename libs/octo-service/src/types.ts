import { RestEndpointMethodTypes } from '@octokit/rest';

export type IssueData =
  RestEndpointMethodTypes['issues']['get']['response']['data'];
export type TPrData =
  RestEndpointMethodTypes['pulls']['get']['response']['data'];

export type TCommitInfo =
  RestEndpointMethodTypes['repos']['getCommit']['response']['data'];

export interface IPrDetail {
  type: 'pr';
  issue: IssueData;
  pr: TPrData;
}
export interface IIssueDetail {
  type: 'issue';
  issue: IssueData;
}

export interface IOrganizationPrResult {
  [login: string]: {
    details: string[];
    total: number;
  };
}

export interface IOrganizationNewContributionsResult {
  [full_name: string]: any;
}
