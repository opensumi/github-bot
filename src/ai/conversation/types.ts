import { ECompletionModel } from '../openai/shared';

export interface IConversationSetting {
  preferredModel?: ECompletionModel;
  apiReverseProxyUrl?: string;
}

export const enum EMessageRole {
  Human = '人类',
  AI = 'AI',
}

export interface IConversationHistoryItem {
  type: EMessageRole;
  str: string;
}

export interface IConversationData {
  data: IConversationHistoryItem[];
}
