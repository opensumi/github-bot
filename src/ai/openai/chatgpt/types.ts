export type Role = 'user' | 'assistant';

export type FetchFn = typeof fetch;

export type SendMessageOptions = {
  conversationId?: string;
  parentMessageId?: string;
  messageId?: string;
  stream?: boolean;
  promptPrefix?: string;
  promptSuffix?: string;
  timeoutMs?: number;
  onProgress?: (partialResponse: ChatMessage) => void;
  abortSignal?: AbortSignal;
};

export type MessageActionType = 'next' | 'variant';

export type SendMessageBrowserOptions = {
  conversationId?: string;
  parentMessageId?: string;
  messageId?: string;
  action?: MessageActionType;
  timeoutMs?: number;
  onProgress?: (partialResponse: ChatMessage) => void;
  abortSignal?: AbortSignal;
};

export interface ChatMessage {
  id: string;
  text: string;
  role: Role;
  parentMessageId?: string;
  conversationId?: string;
  detail?: any;
}

export type ChatGPTErrorType =
  | 'unknown'
  | 'chatgpt:pool:account-on-cooldown'
  | 'chatgpt:pool:account-not-found'
  | 'chatgpt:pool:no-accounts'
  | 'chatgpt:pool:timeout'
  | 'chatgpt:pool:rate-limit'
  | 'chatgpt:pool:unavailable';

export class ChatGPTError extends Error {
  statusCode?: number;
  statusText?: string;
  isFinal?: boolean;
  accountId?: string;
  type?: ChatGPTErrorType;
}

/** Returns a chat message from a store by it's ID (or null if not found). */
export type GetMessageByIdFunction = (id: string) => Promise<ChatMessage>;

/** Upserts a chat message to a store. */
export type UpsertMessageFunction = (message: ChatMessage) => Promise<void>;

/**
 * https://chat.openapi.com/backend-api/conversation
 */
export type ConversationJSONBody = {
  /**
   * The action to take
   */
  action: string;

  /**
   * The ID of the conversation
   */
  conversation_id?: string;

  /**
   * Prompts to provide
   */
  messages: Prompt[];

  /**
   * The model to use
   */
  model: string;

  /**
   * The parent message ID
   */
  parent_message_id: string;
};

export type Prompt = {
  /**
   * The content of the prompt
   */
  content: PromptContent;

  /**
   * The ID of the prompt
   */
  id: string;

  /**
   * The role played in the prompt
   */
  role: Role;
};

export type ContentType = 'text';

export type PromptContent = {
  /**
   * The content type of the prompt
   */
  content_type: ContentType;

  /**
   * The parts to the prompt
   */
  parts: string[];
};

export type ConversationResponseEvent = {
  message?: Message;
  conversation_id?: string;
  error?: string | null;
};

export type Message = {
  id: string;
  content: MessageContent;
  role: Role;
  user: string | null;
  create_time: string | null;
  update_time: string | null;
  end_turn: null;
  weight: number;
  recipient: string;
  metadata: MessageMetadata;
};

export type MessageContent = {
  content_type: string;
  parts: string[];
};

export type MessageMetadata = any;

export type GetAccessTokenFn = ({
  email,
  password,
  sessionToken,
}: {
  email: string;
  password: string;
  sessionToken?: string;
}) => string | Promise<string>;
