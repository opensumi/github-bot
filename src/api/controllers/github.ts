import { webhookHandler } from '@/github';
import { initApp } from '@/github/app';
import { checkTokenValid } from '@/github/octokit/token';
import { GitHubKVManager } from '@/github/storage';

export function route(hono: THono) {
  hono.post('/github/app/:id', async (c) => {
    const id = c.req.param('id') ?? c.req.query('id');
    if (!id) {
      return c.send.error(400, 'need a valid id');
    }
    const githubKVManager = new GitHubKVManager();
    const setting = await githubKVManager.getAppSettingById(id);

    if (!setting) {
      return c.send.error(400, 'id not found in database');
    }
    if (!setting.githubSecret) {
      return c.send.error(400, 'please set app webhook secret in database');
    }

    const app = await initApp(setting);
    return webhookHandler(app.webhooks, c.req, c.executionCtx);
  });

  hono.get('/github/installation-token/:id', async (c) => {
    const id = c.req.param('id');
    const token = c.req.query('token');
    const flag = c.req.query('flag');
    if (!token || !flag) {
      return c.send.error(403, 'invalid request');
    }
    // 如果这个 token 有对应 id 的指定仓库写权限，意味着可以换出来一个安装 token
    const isValid = await checkTokenValid(token);
    if (!isValid) {
      return c.send.error(401, 'invalid token');
    }

    return c.send.message(token);
  });
}
