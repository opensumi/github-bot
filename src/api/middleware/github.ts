import { assert } from '@/api/utils/assert';
import { ValidationError } from '@/github';

import { objectRequired } from '../utils/validator';

export function middleware(hono: THono) {
  hono.use('/github/app/*', async (c, next) => {
    const headers = c.req.headers;

    assert(
      headers.get('User-Agent')?.startsWith('GitHub-Hookshot/'),
      new ValidationError(403, 'User agent: not from GitHub'),
    );

    objectRequired(headers, 'Headers', [
      'x-github-event',
      'x-hub-signature-256',
      'x-github-delivery',
    ]);

    await next();
  });
}
