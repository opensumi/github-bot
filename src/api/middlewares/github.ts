import { z } from 'zod';

import { assert } from '@/api/utils/assert';
import { ValidationError } from '@/github';

const GitHubEvent = z.object({
  'x-github-event': z.string(),
  'x-hub-signature-256': z.string(),
  'x-github-delivery': z.string(),
});

export function middleware(hono: THono) {
  hono.use('/github/*', async (c, next) => {
    const headers = c.req.headers;

    assert(
      headers.get('User-Agent')?.startsWith('GitHub-Hookshot/'),
      new ValidationError(403, 'User agent: not from GitHub'),
    );
    assert(
      headers.get('Content-Type') === 'application/json',
      new ValidationError(415, 'Content type: not json'),
    );
    try {
      GitHubEvent.parse(c.req.headers);
    } catch (error) {
      console.log(`ðŸš€ ~ file: github.ts:28 ~ hono.use ~ error`, error);
      throw error;
    }

    await next();
    c.header('x-message', 'This is middleware!');
  });
}
