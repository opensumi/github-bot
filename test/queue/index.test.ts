import 'dotenv/config';

import { EmitterWebhookEventName } from '@octokit/webhooks';

import { IGitHubEventQueueMessage } from '@/queue/types';
import { GitHubEventWorker } from '@/queue/worker/github';

import { prepareEnv } from '../__mocks__';
import { MockMessageBatch } from '../__mocks__/queue/message';
import {
  pull_request_closed,
  pull_request_opened,
  pull_request_13_opened,
  pull_request_edited_wip,
  pull_request_edited_base,
  issue_opened_event,
  release_published,
  antd_mini_release_published,
  pull_request_review_comment_0_created,
  pull_request_review_4_submitted_changes_requested,
  pull_request_review_comment_1_created,
} from '../fixtures';

const botId = 'mock';

const describe = process.env.GITHUB_TOKEN
  ? global.describe
  : global.describe.skip;

describe('queue', () => {
  beforeAll(() => {
    prepareEnv();
  });

  it('should work', async () => {
    const wk = new GitHubEventWorker('app');
    const events = [
      { name: 'pull_request.closed', payload: pull_request_closed },
      { name: 'pull_request.opened', payload: pull_request_opened },
      { name: 'pull_request.opened', payload: pull_request_13_opened },
      { name: 'pull_request.edited', payload: pull_request_edited_wip },
      { name: 'pull_request.edited', payload: pull_request_edited_base },
      { name: 'issues.opened', payload: issue_opened_event },
      { name: 'release.published', payload: release_published },
      { name: 'release.published', payload: antd_mini_release_published },
      {
        name: 'pull_request_review_comment.created',
        payload: pull_request_review_comment_0_created,
      },
    ] as { name: EmitterWebhookEventName; payload: any }[];

    const githubAppMessage = events.map((v) => ({
      botId,
      type: 'github-app',
      data: {
        id: `${Math.random()}`,
        name: v.name,
        payload: v.payload,
      },
    })) as IGitHubEventQueueMessage[];

    const batch = MockMessageBatch.from(githubAppMessage);
    batch.messages.forEach((v) => {
      wk.push(v);
    });

    await wk.run();
  });

  it('can composite pr review', async () => {
    const wk = new GitHubEventWorker('app');
    const events = [
      {
        name: 'pull_request_review.submitted',
        payload: pull_request_review_4_submitted_changes_requested,
      },
      {
        name: 'pull_request_review_comment.created',
        payload: pull_request_review_comment_0_created,
      },
      {
        name: 'pull_request_review_comment.created',
        payload: pull_request_review_comment_1_created,
      },
    ] as { name: EmitterWebhookEventName; payload: any }[];

    const reviews = events.map((v) => ({
      botId,
      type: 'github-app',
      data: {
        id: `${Math.random()}`,
        name: v.name,
        payload: v.payload,
      },
    })) as IGitHubEventQueueMessage[];

    const batch = MockMessageBatch.from(reviews);
    wk.push(...batch.messages);

    await wk.run();
  });
});
