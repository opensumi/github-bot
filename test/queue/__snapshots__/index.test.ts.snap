// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queue can composite discussion 1`] = `
[
  {
    "eventName": "discussion.created",
    "markdown": {
      "text": "#### [[octo-repo]](https://github.com/octo-org/octo-repo) [Codertocat](https://github.com/Codertocat) created [discussion](https://github.com/octo-org/octo-repo/discussions/90)  
#### [#90 Welcome to discussions!](https://github.com/octo-org/octo-repo/discussions/90)

***

We're glad to have you here!

***

[Codertocat](https://github.com/Codertocat) created [comment](https://github.com/octo-org/octo-repo/discussions/90):  

I have so many questions to ask you!",
      "title": "[octo-repo] Discussion#90 created",
    },
  },
]
`;

exports[`queue can composite pr review 1`] = `
[
  {
    "eventName": "pull_request_review.submitted",
    "markdown": {
      "text": "#### [[Hello-World]](https://github.com/Codertocat/Hello-World) [Codertocat](https://github.com/Codertocat) [requested](https://github.com/Codertocat/Hello-World/pull/2#pullrequestreview-237895671) changes on [pull request](https://github.com/Codertocat/Hello-World/pull/2)  
#### [#2 Update the README with new information.](https://github.com/Codertocat/Hello-World/pull/2)

***

[Codertocat](https://github.com/Codertocat) created [review comment](https://github.com/Codertocat/Hello-World/pull/2#discussion_r284312630):  

Maybe you should use more emoji on this line.

***

[Codertocat](https://github.com/Codertocat) created [review comment](https://github.com/Codertocat/Hello-World/pull/2#discussion_r284312630):  

Maybe you should use more emojji on this line.",
      "title": "[Hello-World] Review requested changes",
    },
  },
]
`;

exports[`queue can composite pr review comment single 1`] = `
[
  {
    "eventName": "pull_request_review_comment.created",
    "markdown": {
      "compactText": "[Codertocat](https://github.com/Codertocat) created [review comment](https://github.com/Codertocat/Hello-World/pull/2#discussion_r284312630):  

Maybe you should use more emojji on this line.",
      "text": "#### [[Hello-World]](https://github.com/Codertocat/Hello-World) [Codertocat](https://github.com/Codertocat) created [review comment](https://github.com/Codertocat/Hello-World/pull/2#discussion_r284312630) on [pull request](https://github.com/Codertocat/Hello-World/pull/2)  
[#2 Update the README with new information.](https://github.com/Codertocat/Hello-World/pull/2)

***

Maybe you should use more emojji on this line.",
      "title": "[Hello-World] Review comment created",
    },
  },
]
`;

exports[`queue can composite pr review single 1`] = `
[
  {
    "eventName": "pull_request_review.submitted",
    "markdown": {
      "compactText": undefined,
      "text": "#### [[core]](https://github.com/opensumi/core) [bytemain](https://github.com/bytemain) [approved](https://github.com/opensumi/core/pull/3345#pullrequestreview-1892322711) [pull request](https://github.com/opensumi/core/pull/3345)  
#### [#3345 feat: support SourceControl.historyProvider](https://github.com/opensumi/core/pull/3345)",
      "title": "[core] Review approved",
    },
  },
]
`;
